lib/core/util.c

```diff
@@ -94,27 +99,39 @@ static struct st_h2o_accept_data_t *create_memcached_accept_data(h2o_accept_ctx_
     return &data->super;
 }
 
-static void setup_accept_data(h2o_accept_ctx_t *ctx, h2o_socket_t *sock, struct timeval connected_at)
+static void free_accept_data(struct st_h2o_accept_data_t *data)
 {
-    struct st_h2o_accept_data_t *data;
+    h2o_timeout_unlink(&data->timeout);
+    free(data);
+}
 
-    if (async_resumption_context.memcached.ctx != NULL) {
-        data = create_memcached_accept_data(ctx, sock, connected_at);
-    } else if (async_resumption_context.redis.host.len != 0) {
-        data = create_redis_accept_data(ctx, sock, connected_at);
-    } else {
-        data = create_accept_data(ctx, sock, connected_at, sizeof(struct st_h2o_accept_data_t));
-    }
+static void free_default_accept_data(struct st_h2o_accept_data_t *_accept_data)
+{
+    free_accept_data(_accept_data);
+}
 
-    sock->data = data;
+static void free_redis_accept_data(struct st_h2o_accept_data_t *_accept_data)
+{
+    struct st_h2o_redis_resumption_accept_data_t *accept_data = (struct st_h2o_redis_resumption_accept_data_t *)_accept_data;
+    assert(accept_data->get_command == NULL);
+    free_accept_data(&accept_data->super);
 }
 
-static void free_accept_data(struct st_h2o_accept_data_t *data)
+static void free_memcached_accept_data(struct st_h2o_accept_data_t *_accept_data)
 {
-    h2o_timeout_unlink(&data->timeout);
-    free(data);
+    struct st_h2o_memcached_resumption_accept_data_t *accept_data = (struct st_h2o_memcached_resumption_accept_data_t *)_accept_data;
+    assert(accept_data->get_req == NULL);
+    free_accept_data(&accept_data->super);
 }
 
+static struct {
+    struct st_h2o_accept_data_t *(*create)(h2o_accept_ctx_t *ctx, h2o_socket_t *sock, struct timeval connected_at);
+    void (*free)(struct st_h2o_accept_data_t *accept_data);
```

Comment:
Please use `destroy` instead of `free`. `free` is a function defined in ISO C, that might be defined as a macro. In such case, the code will fail to compile.

That is the reason we use `destroy` as an antonym to `create` throughout H2O (and FWIW we use `dispose` as an antonym for `init`).